File: backend/server/internal/service/roles.go
package service

const (
	SpaceRoleViewer = "viewer"
	SpaceRoleEditor = "editor"
	SpaceRoleAdmin  = "admin"
)

var spaceRoleRank = map[string]int{
	SpaceRoleViewer: 1,
	SpaceRoleEditor: 2,
	SpaceRoleAdmin:  3,
}

func SpaceRoleRank(name string) int {
	if r, ok := spaceRoleRank[name]; ok {
		return r
	}
	return 0
}

func MaxSpaceRole(roles ...string) string {
	best := ""
	bestRank := 0
	for _, role := range roles {
		if rank := SpaceRoleRank(role); rank > bestRank {
			bestRank = rank
			best = role
		}
	}
	return best
}

// GlobalRolesAsSpaceRole maps platform roles to a space-role floor.
func GlobalRolesAsSpaceRole(globalRoles []string) string {
	if HasRole(globalRoles, "super_admin", "admin") {
		return SpaceRoleAdmin
	}
	if HasRole(globalRoles, "editor") {
		return SpaceRoleEditor
	}
	if HasRole(globalRoles, "viewer") {
		return SpaceRoleViewer
	}
	return ""
}

func HasSpaceRole(effectiveRole, minRole string) bool {
	return SpaceRoleRank(effectiveRole) >= SpaceRoleRank(minRole)
}

func CanEditInSpace(effectiveRole string, globalRoles []string) bool {
	if HasRole(globalRoles, "super_admin") {
		return true
	}
	return HasSpaceRole(effectiveRole, SpaceRoleEditor)
}

func CanUseLLMInSpace(effectiveRole string, globalRoles []string) bool {
	return CanEditInSpace(effectiveRole, globalRoles)
}

func CanManageBooksInSpace(effectiveRole string, globalRoles []string) bool {
	return CanEditInSpace(effectiveRole, globalRoles)
}

File: frontend/src/lib/roles.ts
export type SpaceRole = "viewer" | "editor" | "admin" | "";

interface RoleUser {
  roles: string[];
}

const SPACE_EDITOR_ROLES = new Set<SpaceRole>(["editor", "admin"]);

export function isPlatformAdmin(user: RoleUser | null | undefined): boolean {
  if (!user) return false;
  return user.roles.some((role) => role === "super_admin" || role === "admin");
}

export function canEditInSpace(
  spaceRole: SpaceRole | undefined,
  user: RoleUser | null | undefined,
  canEdit?: boolean,
): boolean {
  if (canEdit === true) return true;
  if (!user) return false;
  if (user.roles.includes("super_admin")) return true;
  return SPACE_EDITOR_ROLES.has(spaceRole ?? "");
}

export function canManageBooksInSpace(
  spaceRole: SpaceRole | undefined,
  user: RoleUser | null | undefined,
  canEdit?: boolean,
): boolean {
  return canEditInSpace(spaceRole, user, canEdit);
}

/** @deprecated Use canManageBooksInSpace with space context */
export function canManageBooks(user: RoleUser | null | undefined): boolean {
  if (!user) return false;
  return user.roles.some((role) => role === "super_admin" || role === "admin" || role === "editor");
}
